### PR TITLE
Feature/alicloud ess scalinggroup vserver groups attachment

### PR DESCRIPTION
**Pull request by TalentNing**
=== RUN   TestAccAlicloudEssVserverGroupsAttachment_basic
--- PASS: TestAccAlicloudEssVserverGroupsAttachment_basic (56.42s)
PASS
ok      github.com/terraform-providers/terraform-provider-alicloud/alicloud     56.475s

### CODE REVIEW

File: website/docs/r/ess_scalinggroup_vserver_groups_attachment.html.markdown [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+-> **NOTE:** Scaling group and vserver groups must be in the same region.
+
+-> **NOTE:** The load balancer of which vserver groups belongs to must be in `active` status.
+
+-> **NOTE:** If scaling group's network type is `VPC`, the vserver groups must be in the same `VPC`.
+ 
+-> **NOTE:** A scaling group can have at most 5 vserver groups attached.
+
+-> **NOTE:** Vserver groups and the default group of loadbalancer share the same backend server quota.
+
```
[the remark this answers is not in the document]
Comment (TalentNing) [reply website/docs/r/ess_scalinggroup_vserver_groups_attachment.html.markdown]: done

File: alicloud/resource_alicloud_ess_scalinggroup_vserver_groups.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+		Delete: resourceAliyunEssVserverGroupsDelete,
+		Schema: map[string]*schema.Schema{
+			"scaling_group_id": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"vserver_groups": {
+				Type:     schema.TypeSet,
+				Optional: true,
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_scalinggroup_vserver_groups.go]: I think the `vserver_groups ` should be required.

File: alicloud/resource_alicloud_ess_scalinggroup_vserver_groups.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	d.Set("scaling_group_id", object.ScalingGroupId)
+	d.Set("vserver_groups", object.VServerGroups.VServerGroup)
+
+	return nil
+}
+
+func resourceAliyunEssVserverGroupsUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	if d.HasChange("vserver_groups") {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_scalinggroup_vserver_groups.go]: The update method have more than one API invoking and please set the partial for it.

File: alicloud/resource_alicloud_ess_scalinggroup_vserver_groups.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+	}
+	attachVserverGroupMap := make(map[string]EssStruct)
+	err = attachOrDetachVserverGroups(d, client, attachVserverGroupMap, detachVserverGroupMap)
+	if err != nil {
+		return WrapError(err)
+	}
+	return nil
+}
+
+func verifyVserverGroups(vserverGroups *schema.Set) (map[string]EssStruct, error) {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_scalinggroup_vserver_groups.go]: I think the method verifyVserverGroups is redundant.

File: alicloud/resource_alicloud_ess_scalinggroup_vserver_groups.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+		})
+		if err != nil {
+			return WrapErrorf(err, DefaultErrorMsg, d.Id(), detachVserverGroupsRequest.GetActionName(), AlibabaCloudSdkGoERROR)
+		}
+		addDebug(detachVserverGroupsRequest.GetActionName(), raw)
+	}
+	return nil
+}
+
+func buildAttachScalingGroupVserverGroupMap(vserverGroupMap map[string]EssStruct) map[string][]ess.AttachVServerGroupsVServerGroupAttribute {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_scalinggroup_vserver_groups.go]: I think the method buildAttachScalingGroupVserverGroupMap also is redundant.

File: alicloud/resource_alicloud_ess_scalinggroup_vserver_groups.go [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+		} else {
+			vserverGroupAttributes := vserverGroupRequestMap[loadbalancerId]
+			vserverGroupAttributes = append(vserverGroupAttributes, vserverGroupAttribute)
+			vserverGroupRequestMap[loadbalancerId] = vserverGroupAttributes
+		}
+	}
+	return vserverGroupRequestMap
+}
+
+func buildDetachScalingGroupVserverGroupMap(vserverGroupMap map[string]EssStruct) map[string][]ess.DetachVServerGroupsVServerGroupAttribute {
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_ess_scalinggroup_vserver_groups.go]: still applies.

File: alicloud/service_alicloud_ess.go [hunk 7]
```diff
@@ -15,6 +15,13 @@ type EssService struct {
 	client *connectivity.AliyunClient
 }
 
+type EssStruct struct {
```
Comment (xiaozhu36) [inline alicloud/service_alicloud_ess.go]: There is no need to use a new struct to record it.

File: website/docs/r/ess_scalinggroup_vserver_groups.markdown [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `scaling_group_id` - (Required) ID of the scaling group.
+* `vserver_groups` - (Optional) The vserver groups attached on scaling group. See [Block vserver_group](#block-vserver_group) below for details.
+
+
+## Attributes Reference
```
Comment (xiaozhu36) [inline website/docs/r/ess_scalinggroup_vserver_groups.markdown]: I think the block `Attributes Reference` should be moved the last and it only has one attribute: `id`

**Pull request merged**